Client/game_sa/CDamageManagerSA.cpp

```diff
@@ -34,50 +34,60 @@ BYTE CDamageManagerSA::GetDoorStatus ( eDoors bDoor )
     return NULL;
 }
 
-VOID CDamageManagerSA::SetDoorStatus ( eDoors bDoor, BYTE bDoorStatus )
+VOID CDamageManagerSA::SetDoorStatus ( eDoors bDoor, BYTE bDoorStatus, bool bFlyingComponent )
 {
     DEBUG_TRACE("VOID CDamageManagerSA::SetDoorStatus ( eDoors bDoor, BYTE bDoorStatus )");
 
     if ( bDoor < MAX_DOORS )
     {
+        unsigned char ucOldStatus = internalInterface->Door[bDoor];
         // Different from before?
-        if ( internalInterface->Door [bDoor] != bDoorStatus )
+        if ( ucOldStatus != bDoorStatus )
         {
+            bool bOldBashed = ucOldStatus == DT_DOOR_BASHED || ucOldStatus == DT_DOOR_BASHED_AND_SWINGING_FREE;
+            bool bNewBashed = bDoorStatus == DT_DOOR_BASHED || bDoorStatus == DT_DOOR_BASHED_AND_SWINGING_FREE;
+
             // Set it
             internalInterface->Door [bDoor] = bDoorStatus;
 
-            // Are we making it intact?
-            if ( bDoorStatus == DT_DOOR_INTACT || bDoorStatus == DT_DOOR_SWINGING_FREE )
+            // Update door model (only if needed - this gonna shut the door)
+            if ( bOldBashed != bNewBashed || bDoorStatus == DT_DOOR_MISSING || ucOldStatus == DT_DOOR_MISSING )

```

Comment:
Fixes issue 2 & 3. Updating door model causes angle reset so we want to avoid this in some situations (for example changing from state 0 to 1).